### ISSUE TITLE
Hyperlinkable messages

### ISSUE THREAD

**Issue by dom96** [nim-lang/nimbot#4]
add ability to get a link to a specific message in the chat logs.

**Comment by ddmgy** [comment]
I have a couple of questions about how you would like this to look and work. I assume it would be something like how you can link to lines in files on Github (https://github.com/nim-lang/Nim/blob/devel/lib/system.nim#L166), but what would be the equivalent of a line in this? Would you be able to link only to messages, or would you be able to link to everything in the log, including joins, quits, nick changes, etc.? When you have the link to a specific line/message, would it need to be highlighted? I don't know if highlighting would be feasible without Javascript magic, though.

Also, would you click on the date in the log to get the link to that line, similar to clicking the line number on Github?  I think, when rendering the log, it would be possible to wrap the date part in an `a` tag with an id and href attribute corresponding to whatever the line number is. I'll have to play around with the htmlgen module.

Let me know if you have any ideas of how you would like these links to act, and I'll take a crack at it this weekend.

**Comment by dom96** [comment]
> I assume it would be something like how you can link to lines in files on Github

Precisely!

> but what would be the equivalent of a line in this?

Every line in the IRC logs would be a linkable line, this includes joins, quits, nick changed etc.

> When you have the link to a specific line/message, would it need to be highlighted?

Yes, if possible. Pretty sure Github manages to do it with no JS.

> Also, would you click on the date in the log to get the link to that line, similar to clicking the line number on Github?

Well, you would click on the timestamp but yeah.

> Let me know if you have any ideas of how you would like these links to act, and I'll take a crack at it this weekend.

A good approach would be to simply copy how Github does it. Just s/line number/timestamp/. :)

**Comment by ddmgy** [comment]
So, I've got hyperlinks working in the logs. Click on the date/timestamp on the left, get link to that line. It worked just as I thought it would. However, highlighting that line may be a little more difficult. Unless I'm completely missing something while inspecting the source in a Github page, it does use Javascript to dynamically change the style on the line you click.

For instance, if I were to click the line number 73 in https://github.com/nim-lang/nimbot/blob/master/src/irclogrender.nim, it navigates the page to that line, and adds a style="background-color: ..." attribute to that line in the file. If I then click on a different line, it removes the style attribute from line 73, and instead adds it to the newly clicked line.

One way that I was thinking could work that does not require javascript, is to capture the linked line in the jester route (like in https://github.com/nim-lang/nimbot/blob/master/src/nimbot.nim#L244). Then that could be passed to irclogrender.renderHtml, then to renderItems, where the item whose timestamp matches would have a style applied to it. I don't really like it, seems like a kludge, but I'd like to hear any thoughts you have about it.

Another way would be to add a css rule to log.css, then add that element to the selected line in the log. I can't test this on my machine, however, because the stylesheets cannot be found (404) when I run nimbot, no matter where I put the stylesheets in relation to the nimbot binary. So, my irc logs look like [this](http://i.imgur.com/W0Y2Egz.jpg), and the official logs look like [this](http://i.imgur.com/9RVxgyI.jpg). It's not a big deal, but it would be difficult to test unless I figure out a way past this.

Also, how would you like the URL fragment to look? The bit after the hash mark. Github uses #Ln, where n is just an integer, and that works when I try it in the logs. I've also just used the formatted timestamp (#00:04:32), and that works as well. The timestamp has the benefit of being more clear, since the logs don't actually have any line numbers. Either way would be easy to add.

**Comment by ddmgy** [comment]
Okay, I've cleaned up the changes I made to allow links to messages in the irc logs and pushed them. Here is the commit: https://github.com/ddmgy/nimbot/commit/c417c69afdf56f9143b57a370a550f799417b1a3.

I did not open a pull request yet because I still want to take a look at highlighting the linked message. I'm having trouble capturing it in the jester route, and I am unaware of another way to get the URL fragment in Nim.

Here's how today's logs look with the timestamp links: http://i.imgur.com/ZcSGmoZ.jpg (I still haven't figured out how to fix the 404 error on getting the stylesheets).

If I cannot figure out how to get the linked message to get the highlighting to work, I'll make a PR soon.

**Comment by dom96** [comment]
Looks good. Use JS if you must. You can get the stylesheet working by
running nimbot in the src directory. Jester looks for static files in
./public.

**Comment by ddmgy** [comment]
Alright, thanks for the tip! Now my nimbot doesn't look terrible.

I've got highlighting working! It took a surprisingly small amount of Javascript and CSS, and it works on all browsers I have available. I tested it on:
- Firefox 45 and Chrome 49 for Android (4.4, 4.5, 5, 6)
- Chrome 35 and latest Safari  for iOS 6
- Firefox 45, Chromium 49, and Opera 36 on Arch Linux (64-bit)
- Firefox 43, Chrome 45, and Edge 25 on Windows 10 (64-bit)

Except for Safari on Mac OS X (to which I do not have access), this seems like a fairly decent coverage for testing. It should work with any normal browser from within the last ~6 years, anyway.

I've uploaded a screen recording to show the links and highlighting [here](http://webm.host/393a8/). Also, I can minify [(like this)](https://gist.github.com/ddmgy/92eebe601c54817003dd) the javascript I used to save about 100 characters (all whitespace). If you don't mind having a difficult-to-read file in the repo, I can replace the existing version to save some space.

If it all looks good, I can open a PR tomorrow evening.
